src/main/scala/com/etsy/conjecture/scalding/util/HyperparameterSearcher.scala

```diff
@@ -0,0 +1,186 @@
+       }.project('trial, 'instances)
+  }
+
+  def trainTrials(instances: Pipe): Pipe = {
+       instances
+         .mapTo(('trial, 'instances) -> ('trial, 'model)) {
+           x: (Int, List[LabeledInstance[L]]) =>
+           //In the unlike case that a setting does not exist, use the default value
+           val args: Args = settings.getOrElse(x._1, option.unParse)
+           val instanceSet = x._2
+           val modelTrainer = getModelTrainer(args)
+           val model = modelTrainer.getModel
+           //Train model
+           instanceSet.foreach(model.update)
+           (x._1, model)
+        }
+  }
+
+  def evaluate (models: Pipe, testSet: Pipe): Pipe = {
+     val eval = models
+       .joinWithSmaller('trial -> 'trial, testSet)
+       .mapTo(('model, 'instances, 'trial) -> ('eval, 'setting, 'model)) {
+          x: (M, List[LabeledInstance[L]], Int) =>
+          val model = x._1
+          val testList = x._2
+          val args = settings.getOrElse(x._3, option.unParse)
+          (evaluateAccuracy(testList, model),  args.toString, model)
+       }
+       eval
+  }
+
+  def evaluateAccuracy(instances: List[LabeledInstance[L]], model: M): Double = {
+        val eval = evaluator.build
+        instances.map {
+            instance: LabeledInstance[L] =>
+                val realLabel = instance.getLabel
+                val prediction = model.predict(instance.getVector)
+                eval.add(realLabel, prediction)
+        }
+        val agg = new EvaluationAggregator[L]()
+        agg.add(eval)
+        agg.getValue("Acc (avg)")
+    }
+}
+
+
+class BinaryHyperparameterSearcher(option: DynamicOptions, parameters: Seq[HyperParameter[_]],
+                                 numTrials: Int) extends HyperparameterSearcher[BinaryLabel, UpdateableLinearModel[BinaryLabel], BinaryModelEvaluation](option, parameters, numTrials) {
+    val evaluator = new BinaryEvaluator()
+    def getModelTrainer(args: Args) = new BinaryModelTrainer(args)
+}
+
+class RegressionHyperparameterSearcher(option: DynamicOptions, parameters: Seq[HyperParameter[_]],
+                                 numTrials: Int) extends HyperparameterSearcher[RealValuedLabel, UpdateableLinearModel[RealValuedLabel], RegressionModelEvaluation](option, parameters, numTrials) {
+    val evaluator = new RegressionEvaluator()
+    def getModelTrainer(args: Args) = new RegressionModelTrainer(args)
+}
+
+class MulticlassHyperparameterSearcher(option: DynamicOptions, parameters: Seq[HyperParameter[_]], numTrials: Int, categories: Array[String]) extends HyperparameterSearcher[MulticlassLabel, UpdateableMulticlassLinearModel, MulticlassModelEvaluation](option, parameters, numTrials) {
+    val evaluator = new MulticlassEvaluator(categories)
+    def getModelTrainer(args: Args) = new MulticlassModelTrainer(args, categories)
+}
+
+trait ParameterSampler[T] extends Serializable {
+  def sample(rng: scala.util.Random): T
+  def buckets: Array[(T,Double,Double,Double,Int)]
```

Comment:
can you put in some comments aout what this is doing and what the stuff in the array is